keps/sig-windows/2802-identify-windows-pods-apiserver-admission/README.md

```diff
@@ -0,0 +1,344 @@
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The main goal of this enhancement is to authoritatively identify Windows
+pods during the API Server admission time. 
+
+
+## Motivation
+
+Identifying Windows pods during the API Server admission time is crucial to apply
+appropriate security constraints to the pod. Without proper identification, some admission
+plugins may apply unnecessary security constraints to the Windows pods or in the
+worst case, don't apply security constraints at all.
+
+
+### Goals
+
+Admission plugins and/or Admission Webhooks identifying if the incoming pod
+is Windows or not.
+
+
+### Non-Goals
+
+
+- Validating if the Windows pods have appropriate security constraints.
+- Interaction with container runtimes.
+
+## Proposal
+
+
+As mentioned [earlier](#Motivation) identifying the Windows pods during the scheduling phase can be done
+in the following ways:
+- Based on [nodeSelector and tolerations](https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-containers/#ensuring-os-specific-workloads-land-on-the-appropriate-container-host) in the pod spec with 
+  Windows node specific labels.
+- Based on [runtimeclasses](https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-containers/#ensuring-os-specific-workloads-land-on-the-appropriate-container-host) in the pod spec
+
+The problem with using nodeSelector and tolerations is that any unprivileged user/entity can apply the 
+nodeSelector and/or tolerations in the pod spec making it unsecure whereas RuntimeClasses are [recommended](https://kubernetes.io/docs/concepts/containers/runtime-class/#setup) to be
+created by cluster administrator making them authoritative enough to be used.
+
+### User Stories (Optional)
+
+
+#### Story 1
+
+As a Kubernetes cluster administrator, I want appropriate security contexts to be 
+applied to my Windows Pods along with Linux pods
+
+#### Story 2
+
+As a Kubernetes cluster administrator, I want to use my own admission webhook for
+Windows pods.
+
+### Notes/Constraints/Caveats (Optional)
```

Comment:
We should note that operators/users would be forced to use runtime classes in all of their podspecs (or at least for all the ones targeting Windows nodes/containers) we should call that out here.

Reply:
And also define what would happen if someone tries to schedule a pod with just a`kubernetes.io\os=windows` nodeSelector (with no runtime class specified)

Reply:
> We should note that operators/users would be forced to use runtime classes in all of their podspecs (or at least for all the ones targeting Windows nodes/containers) we should call that out here.

Not in the initial release but in the subsequent releases.

> And also define what would happen if someone tries to schedule a pod with just akubernetes.io\os=windows nodeSelector (with no runtime class specified)

We'd emit a warning just like how deprecated api usage would cause warning in first release. In the subsequent releases, we'd deny pod admission if it just has `kubernetes.io/os=Windows` nodeSelector as mentioned [here](https://github.com/kubernetes/enhancements/pull/2803/files#diff-0bd86b6b4704c4e10a453c7f15e74e31aacb53891087cf069ae11b8bf5293a30R172)

Reply:
typo: `kubernetes.io/os=windows`

> In the subsequent releases, we'd deny pod admission if it just has `kubernetes.io/os=Windows` nodeSelector as mentioned [here](https://github.com/kubernetes/enhancements/pull/2803/files#diff-0bd86b6b4704c4e10a453c7f15e74e31aacb53891087cf069ae11b8bf5293a30R172)

that seems like a significant compatibility break... I don't anticipate pods like that getting rejected in the future